pretixprint/app/src/main/java/eu/pretix/pretixprint/byteprotocols/BrotherRaster.kt

```diff
@@ -0,0 +1,193 @@
+        if (label.continuous) {
+            ostream.write(byteArrayOf(0x1B, 'i'.code.toByte(), 'd'.code.toByte(), 0x23, 0x00)) // Margins: 3mm
+        } else {
+            ostream.write(byteArrayOf(0x1B, 'i'.code.toByte(), 'd'.code.toByte(), 0x00, 0x00)) // Margins: 0mm
+        }
+
+        val pixels = IntArray(scaled.width * scaled.height)
+        scaled.getPixels(pixels, 0, scaled.width, 0, 0, scaled.width, scaled.height)
+        val bytewidth = scaled.width / 8
+        for (y in 0 until maxHeight) {
+            val row = ByteArray(bytewidth)
+            for (xoffset in 0 until bytewidth) {
+                var col = 0
+                // check for overprinting (more height than scaled image)
+                if (y <= scaled.height) {
+                    for (j in 0..7) {
+                        val px = pixels[min((xoffset * 8 + j) + scaled.width * y, pixels.size - 1)]
+                        if ((px shr 24) and 0xff > 128 && ((px shr 16) and 0xff < 128 || (px shr 8) and 0xff < 128 || px and 0xff < 128)) {
+                            // A > 128 && (R < 128 || G < 128 || B < 128)
+                            col = col or (1 shl (7 - j))
+                        }
+                    }
+                } else {
+                    // fill rest of paper with white nothingness
+                    col = 0
+                }
+                row[xoffset] = col.toByte()
+            }
+
+            ostream.write(if (label.twoColor) 'w'.code else 'g'.code)
+            ostream.write(if (label.twoColor) 0x01 else 0x00) // default color
+            val len = min(90, row.size)
+            ostream.write(len)
+            ostream.write(row, 0, len)
+            ostream.flush()
+
+            // HACK for printing usual black/white on red/black labels
+            if (label.twoColor) {
+                ostream.write('w'.code)
+                ostream.write(0x02) // second color
+                ostream.write(90)
+                ostream.write(ByteArray(90), 0, 90) // simply nothing
+                ostream.flush()
+            }
+        }
+
+        // Print command with feeding
+        if (isLastPage) {
+            ostream.write(0x1A)
+        } else {
+            ostream.write(0x0C)
+        }
+
+        return ostream.toByteArray()
+    }
+
+    override fun send(pages: List<CompletableFuture<ByteArray>>, istream: InputStream, ostream: OutputStream, conf: Map<String, String>, type: String) {
+        // Invalidate
+        for (i in 0 until 400) {
+            ostream.write(0x00)
+        }
+        ostream.flush()
+
+        // Initialize
+        ostream.write(byteArrayOf(0x1B, '@'.code.toByte()))
+
+        for (f in pages) {
+            ostream.write(f.get())
+            ostream.flush()
+        }
+
+        val wap = Integer.valueOf(conf.get("hardware_${type}printer_waitafterpage") ?: "2000")
+        Thread.sleep(wap.toLong())
```

Comment:
This is wrong, and it's also wrong in GraphicESCPOS where you copied it from. Sorry :D

We have two different types of waiting implemented:

* FGL and SLCS perform a `Thread.sleep(2000)` at the **end of the job**. This is because we experienced trouble with some printers if you close the TCP socket too quickly. However, there is almost no impact on the user since our background thread just sleeps for 2 more seconds before it shuts itself down, but it does not slow down printing, so no one cares. In this case, we should just hardcode the 2000ms

* ESCPOS performs a *configurable* `Thread.sleep` **after every page**. This is because some ESCPOS printers have a very small buffer and if we send them 4 pages at once, they will only print 2. To make it worse, the required delay appears to be dependent on the connection (e.g. Bluetooth is worse than Ethernet), so we've made it configurable.

GraphicESCPOS should probably behave just like ESCPOS, the sleep should be within the page loop.

And for Brother, can you do a few tests with e.g. a 5-or-more-ticket order in pretixPOS and see if it works without a sleep between pages? If yes, we should do the hardcoded sleep like FGL and SLCS. If not, we need the configurable sleep.

Reply:
Just tested it with printing six tickets in one order, the printer has no problem with too much data or early tcp socket closing. Hardcoding 2000ms at the end of the job seems okay, i'll drop the setting.

Interestingly, for printing over an USB connection [they describe some waiting time for cooling](https://download.brother.com/welcome/docp100278/cv_ql800_eng_raster_100.pdf#page=42) - but this should happen while sending raster data 🤔